barretenberg/cpp/src/barretenberg/boomerang_value_detection/graph.cpp

```diff
@@ -0,0 +1,672 @@
+                                                                    size_t gate_index)
+{
+
+    auto to_real = [&](uint32_t variable_index) { return ultra_circuit_builder.real_variable_index[variable_index]; };
+    auto find_position = [&](uint32_t variable_index) {
+        return variables_in_one_gate.contains(to_real(variable_index));
+    };
+    auto& lookup_block = ultra_circuit_builder.blocks.lookup;
+    std::unordered_set<BasicTableId> sha256_plookup_tables{ BasicTableId::SHA256_WITNESS_SLICE_3,
+                                                            BasicTableId::SHA256_WITNESS_SLICE_7_ROTATE_4,
+                                                            BasicTableId::SHA256_WITNESS_SLICE_8_ROTATE_7,
+                                                            BasicTableId::SHA256_WITNESS_SLICE_14_ROTATE_1,
+                                                            BasicTableId::SHA256_BASE16,
+                                                            BasicTableId::SHA256_BASE16_ROTATE2,
+                                                            BasicTableId::SHA256_BASE16_ROTATE6,
+                                                            BasicTableId::SHA256_BASE16_ROTATE7,
+                                                            BasicTableId::SHA256_BASE16_ROTATE8,
+                                                            BasicTableId::SHA256_BASE28,
+                                                            BasicTableId::SHA256_BASE28_ROTATE3,
+                                                            BasicTableId::SHA256_BASE28_ROTATE6 };
+    if (sha256_plookup_tables.contains(table_id)) {
+        uint32_t real_right_idx = to_real(lookup_block.w_r()[gate_index]);
+        uint32_t real_out_idx = to_real(lookup_block.w_o()[gate_index]);
+        if (variables_gate_counts[real_out_idx] != 1 || variables_gate_counts[real_right_idx] != 1) {
+            auto q_m = lookup_block.q_m()[gate_index];
+            auto q_c = lookup_block.q_c()[gate_index];
+            bool find_out = find_position(real_out_idx);
+            bool find_right = find_position(real_right_idx);
+            if (q_c == 0) {
+                if (find_out) {
+                    variables_in_one_gate.erase(real_out_idx);
+                }
+            }
+            if (q_m == 0) {
+                if (find_right) {
+                    variables_in_one_gate.erase(real_right_idx);
+                }
+            }
+            if (table_id == SHA256_BASE16_ROTATE2 || table_id == SHA256_BASE28_ROTATE6) {
+                // we want to remove false cases for special tables even though their selectors != 0
+                // because it's from_1_to_2_table cases, and they aren't dangerous
+                variables_in_one_gate.erase(real_out_idx);
+            }
+        }
+    }
+}
+
+template <typename FF>
+inline void Graph_<FF>::process_current_plookup_gate(bb::UltraCircuitBuilder& ultra_circuit_builder,
```

Comment:
Since it's referenced earlier, it would be better to have it before `remove_unnecessary_plookup_variables`